Git unit tests failing on Ubuntu20.04

### Steps to reproduce

Bisected to https://github.com/spack/spack/pull/51813 see https://github.com/spack/spack/pull/51813#issuecomment-3922932695

To reproduce the issue pull this Docker image:
```console
docker run -it --rm ghcr.io/spack/ubuntu-focal:develop
```
Then run:
```console
apt update && apt install -y python3.8-venv
python3 -m venv ./.venv/py3
. .venv/py3/bin/activate
pip install pytest
export SPACK_PYTHON=/root/.venv/py3/bin/python
spack cd -r
pytest lib/spack/spack/test/git_fetch.py::test_get_full_repo
```

### Error message

You'll get the following output
```console
(py3) root@b223b9781cc8:/opt/spack# pytest lib/spack/spack/test/git_fetch.py::test_get_full_repo
============================================================================================================================== test session starts ==============================================================================================================================
platform linux -- Python 3.8.10, pytest-8.3.5, pluggy-1.5.0
rootdir: /opt/spack
configfile: pytest.ini
collected 24 items                                                                                                                                                                                                                                                              

lib/spack/spack/test/git_fetch.py .FE......ssssssssssssssss                                                                                                                                                                                                               [100%]

==================================================================================================================================== ERRORS =====================================================================================================================================
___________________________________________________________________________________________________________ ERROR at teardown of test_get_full_repo[None-True-False] ____________________________________________________________________________________________________________

request = <SubRequest 'check_for_leftover_stage_files' for <Function test_get_full_repo[None-True-False]>>, mock_stage = '/tmp/pytest-of-root/pytest-0/mock-stage0'
ignore_stage_files = {'.lock', 'build_cache', 'spack-src', 'spack-stage-git-test-git-2m3l3g4zjbfomwsvjbb77kyyzuwzkb7y'}

    @pytest.fixture(scope="function", autouse=True)
    def check_for_leftover_stage_files(request, mock_stage, ignore_stage_files):
        """
        Ensure that each (mock_stage) test leaves a clean stage when done.
    
        Tests that are expected to dirty the stage can disable the check by
        adding::
    
            @pytest.mark.disable_clean_stage_check
    
        and the associated stage files will be removed.
        """
        yield
    
        if mock_stage is None:
            # When tests opt out with @pytest.mark.nomockstage, do not check for left-over files
            return
    
        files_in_stage = set()
        try:
            stage_files = os.listdir(mock_stage)
            files_in_stage = set(stage_files) - ignore_stage_files
        except OSError as err:
            if err.errno == errno.ENOENT or err.errno == errno.EINVAL:
                pass
            else:
                raise
    
        if "disable_clean_stage_check" in request.keywords:
            # clean up after tests that are expected to be dirty
            for f in files_in_stage:
                path = os.path.join(mock_stage, f)
                remove_whatever_it_is(path)
        else:
            ignore_stage_files |= files_in_stage
>           assert not files_in_stage
E           AssertionError: assert not {'spack-stage-git-test-git-2m3l3g4zjbfomwsvjbb77kyyzuwzkb7y'}

lib/spack/spack/test/conftest.py:601: AssertionError
----------------------------------------------------------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------------------------------------------------------
Initialized empty Git repository in /tmp/tmpkl73exwo/git-test/.git/
=================================================================================================================================== FAILURES ====================================================================================================================================
______________________________________________________________________________________________________________________ test_get_full_repo[None-True-False] ______________________________________________________________________________________________________________________

get_full_repo = False, use_commit = True, git_version = Version("2.25.1"), mock_git_repository = <spack.util.pattern.Bunch object at 0x7f8d89817520>, default_mock_concretization = <function default_mock_concretization.<locals>._func at 0x7f8d8960a940>
mutable_mock_repo = <spack.repo.RepoPath object at 0x7f8d896f4fd0>, monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f8d898174f0>

    @pytest.mark.parametrize("get_full_repo", [True, False])
    @pytest.mark.parametrize("use_commit", [True, False])
    def test_get_full_repo(
        get_full_repo,
        use_commit,
        git_version,
        mock_git_repository,
        default_mock_concretization,
        mutable_mock_repo,
        monkeypatch,
    ):
        """Ensure that we can clone a full repository."""
    
        if git_version < Version(min_opt_string):
            pytest.skip("Not testing get_full_repo for older git {0}".format(git_version))
    
        # newer git allows for direct commit fetching
        can_use_direct_commit = git_version >= Version(min_direct_commit)
    
        secure = True
        type_of_test = "tag-branch"
    
        t = mock_git_repository.checks[type_of_test]
    
        spec_string = "git-test"
    
        s = default_mock_concretization(spec_string)
    
        args = copy.copy(t.args)
        args["get_full_repo"] = get_full_repo
        monkeypatch.setitem(s.package.versions, Version("git"), args)
    
        if use_commit:
            git_exe = mock_git_repository.git_exe
            url = mock_git_repository.url
            commit = git_exe("ls-remote", url, t.revision, output=str).strip().split()[0]
            s.variants["commit"] = SingleValuedVariant("commit", commit)
    
        with s.package.stage:
            with spack.config.override("config:verify_ssl", secure):
                s.package.do_stage()
                with working_dir(s.package.stage.source_path):
                    branches = mock_git_repository.git_exe("branch", "-a", output=str).splitlines()
                    nbranches = len(branches)
                    commits = mock_git_repository.git_exe(
                        "log",
                        "--graph",
                        "--pretty=format:%h -%d %s (%ci) <%an>",
                        "--abbrev-commit",
                        output=str,
                    ).splitlines()
                    ncommits = len(commits)
    
            if get_full_repo:
                # default branch commit, plus checkout commit
                assert ncommits == 2, commits
                assert nbranches >= 5, branches
            else:
                assert ncommits == 1, commits
                if can_use_direct_commit:
                    if use_commit:
                        # only commit (detached state)
>                       assert nbranches == 1, branches
E                       AssertionError: ['* (HEAD detached at 8dbba58)', '  tag-branch', '  remotes/origin/tag-branch']
E                       assert 3 == 1

lib/spack/spack/test/git_fetch.py:309: AssertionError
----------------------------------------------------------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------------------------------------------------------
Initialized empty Git repository in /tmp/tmpkl73exwo/git-test/.git/
============================================================================================================================= slowest 30 durations ==============================================================================================================================
1.02s call     lib/spack/spack/test/git_fetch.py::test_get_full_repo[None-True-True]
0.45s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[None-True-True]
0.29s call     lib/spack/spack/test/git_fetch.py::test_get_full_repo[None-True-False]
0.27s call     lib/spack/spack/test/git_fetch.py::test_get_full_repo[None-False-False]
0.26s call     lib/spack/spack/test/git_fetch.py::test_get_full_repo[None-False-True]
0.26s call     lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.2-True-False]
0.26s call     lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.2-True-True]
0.25s call     lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.2-False-True]
0.25s call     lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.2-False-False]
0.10s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.2-True-False]
0.07s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[None-True-False]
0.07s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.2-False-True]
0.07s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[None-False-False]
0.07s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.2-False-False]
0.07s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.2-True-True]
0.07s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.1-True-True]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[None-False-True]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.7.10-True-True]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.7.10-True-False]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.7.0-False-True]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.7.10-False-False]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.7.1-False-False]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.7.0-True-True]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.1-False-True]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.7.0-False-False]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.7.1-True-True]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.7.1-False-True]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.7.1-True-False]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.1-False-False]
0.04s setup    lib/spack/spack/test/git_fetch.py::test_get_full_repo[1.8.5.1-True-False]
============================================================================================================================ short test summary info ============================================================================================================================
SKIPPED [4] lib/spack/spack/test/git_fetch.py:261: Not testing get_full_repo for older git 1.8.5.1
SKIPPED [4] lib/spack/spack/test/git_fetch.py:261: Not testing get_full_repo for older git 1.7.10
SKIPPED [4] lib/spack/spack/test/git_fetch.py:261: Not testing get_full_repo for older git 1.7.1
SKIPPED [4] lib/spack/spack/test/git_fetch.py:261: Not testing get_full_repo for older git 1.7.0
ERROR lib/spack/spack/test/git_fetch.py::test_get_full_repo[None-True-False] - AssertionError: assert not {'spack-stage-git-test-git-2m3l3g4zjbfomwsvjbb77kyyzuwzkb7y'}
FAILED lib/spack/spack/test/git_fetch.py::test_get_full_repo[None-True-False] - AssertionError: ['* (HEAD detached at 8dbba58)', '  tag-branch', '  remotes/origin/tag-branch']
=============================================================================================================== 1 failed, 7 passed, 16 skipped, 1 error in 4.69s ================================================================================================================

```

### Information on your system

* **Spack:** 1.2.0.dev0 (https://github.com/spack/spack/commit/124b9686d153a7e6d6eb9eed7fb318d7fc65272f)
* **Builtin repo:** https://github.com/spack/spack-packages/commit/2659fa4a2cb639664f6932a8cccb7fe7894056c6
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

--- a/lib/spack/spack/test/git_fetch.py
+++ b/lib/spack/spack/test/git_fetch.py
@@ -283,7 +283,7 @@ def test_get_full_repo(
         s.variants["commit"] = SingleValuedVariant("commit", commit)
         if can_use_direct_commit:
             path = mock_git_repository.path
-            git_exe("-C", path, "config", "set", "uploadpack.allowReachableSHA1InWant", "true")
+            git_exe("-C", path, "config", "uploadpack.allowReachableSHA1InWant", "true")
 
     with s.package.stage:
         with spack.config.override("config:verify_ssl", secure):

--- a/lib/spack/spack/test/git_fetch.py
+++ b/lib/spack/spack/test/git_fetch.py
@@ -281,6 +281,9 @@ def test_get_full_repo(
         url = mock_git_repository.url
         commit = git_exe("ls-remote", url, t.revision, output=str).strip().split()[0]
         s.variants["commit"] = SingleValuedVariant("commit", commit)
+        if can_use_direct_commit:
+            path = mock_git_repository.path
+            git_exe("-C", path, "config", "set", "uploadpack.allowReachableSHA1InWant", "true")
 
     with s.package.stage:
         with spack.config.override("config:verify_ssl", secure):

--- a/lib/spack/spack/util/git.py
+++ b/lib/spack/spack/util/git.py
@@ -324,8 +324,9 @@ def git_init_fetch(url, ref, depth=None, debug=False, dest=None, git_exe=None):
 
     filter_args = FILTER_BLOB_NONE(version)
     if filter_args:
-        fetch.extend([*filter_args, url, ref])
+        fetch.extend(filter_args)
         cmds.insert(1, ["config", "extensions.partialClone", "true"])
+    fetch.extend([url, ref])
     _exec_git_commands_unique_dir(git_exe, cmds, debug, dest)
 
 

--- a/lib/spack/spack/util/git.py
+++ b/lib/spack/spack/util/git.py
@@ -315,14 +315,17 @@ def git_init_fetch(url, ref, depth=None, debug=False, dest=None, git_exe=None):
     remote = ["remote", "add", "origin", url]
     config = ["config", "remote.origin.fetch", "+refs/heads/*:origin/refs/*"]
     fetch = ["fetch"]
+    cmds = [init, remote, config, fetch]
 
     if not debug:
         fetch.append("--quiet")
     if depth and protocol_supports_shallow_clone(url):
         fetch.extend(DEPTH(version, str(depth)))
 
-    fetch.extend([*FILTER_BLOB_NONE(version), url, ref])
-    cmds = [init, remote, config, fetch]
+    filter_args = FILTER_BLOB_NONE(version)
+    if filter_args:
+        fetch.extend([*filter_args, url, ref])
+        cmds.insert(1, ["config", "extensions.partialClone", "true"])
     _exec_git_commands_unique_dir(git_exe, cmds, debug, dest)
 
 

--- a/lib/spack/spack/util/git.py
+++ b/lib/spack/spack/util/git.py
@@ -315,7 +315,6 @@ def git_init_fetch(url, ref, depth=None, debug=False, dest=None, git_exe=None):
     remote = ["remote", "add", "origin", url]
     config = ["config", "remote.origin.fetch", "+refs/heads/*:origin/refs/*"]
     fetch = ["fetch"]
-    cmds = [init, remote, config, fetch]
 
     if not debug:
         fetch.append("--quiet")
@@ -325,8 +324,13 @@ def git_init_fetch(url, ref, depth=None, debug=False, dest=None, git_exe=None):
     filter_args = FILTER_BLOB_NONE(version)
     if filter_args:
         fetch.extend(filter_args)
-        cmds.insert(1, ["config", "extensions.partialClone", "true"])
     fetch.extend([url, ref])
+
+    partial_clone = ["config", "extensions.partialClone", "true"] if filter_args else None
+    if partial_clone is not None:
+        cmds = [init, partial_clone, remote, config, fetch]
+    else:
+        cmds = [init, remote, config, fetch]
     _exec_git_commands_unique_dir(git_exe, cmds, debug, dest)
 
 

